aten/src/ATen/native/cuda/AdaptiveAveragePooling.cu

```diff
@@ -220,46 +438,115 @@ namespace {
         "empty");
     }
 
-    TORCH_CHECK((input.ndimension() == 3 || input.ndimension() == 4),
-      "non-empty 3D or 4D (batch mode) tensor expected for input");
     Tensor input_ = input;
-    int64_t grid_x = input.size(-3);
-    if (input.ndimension() == 4) {
-       input_ = input.contiguous();
-       grid_x *= input_.size(-4);
-    }
-    int64_t sizeD  = input_.size(-3);
-    int64_t isizeH = input_.size(-2);
-    int64_t isizeW = input_.size(-1);
-
-    int64_t istrideD = input_.stride(-3);
-    int64_t istrideH = input_.stride(-2);
-    int64_t istrideW = input_.stride(-1);
-
-    int64_t osizeH = output_size[0];
-    int64_t osizeW = output_size[1];
-    if (input.ndimension() == 4) {
-       output.resize_({input_.size(-4), sizeD, osizeH, osizeW});
+    if (input.suggest_memory_format() == at::MemoryFormat::ChannelsLast) {
```

Comment:
Please replace with switch https://github.com/pytorch/pytorch/wiki/Writing-memory-format-aware-operators
and move each version to separate function (as this nesting getting hard to read and maintain).